### DISCUSSION TITLE
ERROR - LocalTaskJob heartbeat got an exception

### DISCUSSION THREAD

**Discussion by kavya-5** [apache/airflow#23600] [General] [answered]
### Apache Airflow version

2.2.3

### What happened

Hi, I am running a DAG in airflow that takes custom params as input.
This param is passed to a PythonOperator to be used in the logic.
This PythonOperator is expected to run for at least 12 hours until a condition is met. But, there is an error encountered in within few mins maybe about 20 mins from the start of the task (PythonOperator) that is `ERROR - LocalTaskJob heartbeat got an exception` 
There is also another exception : `sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) connection to server at "airflow-sqlproxy-service.default.svc.cluster.local" (10.67.240.74), port 3306 failed: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.`

### What you think should happen instead

Before using the custom params, these PythonOperators would run for 12 hours and execute the tasks successfully.
This ERROR occurred after switching to params.

### How to reproduce

To reproduce this issue, maybe one can excute a PythonOperatro as follows

```
with models.DAG(
        'Test_dag',
        schedule_interval='0 0 * * *',
        start_date=dt.datetime(2022,4,14),
        default_args={
            'depends_on_past': False,
            'retries': 10,
            'retry_delay': timedelta(minutes=10)
            },
        params={
        "date": date.today().strftime("%Y%m%d")
    },
        catchup=False) as dag:
    checktime = PythonOperator(
    task_id= "Check_time",
    python_callable= checkTime,
    op_kwargs={
            "date": "{{params.date}}",
           },
    dag=dag,
    )
    def checkTime(date):
       today=dt.datetime.strptime(date, "%Y%m%d")
       runtime=today.replace(hour=12, minute=0, second=0, microsecond=0)
        import time
        while True:

            current_time = dt.datetime.now()
            if current_time>=runtime:
                break
            else:
                time.sleep(45)
```

### Operating System

I am running Airflow on Google Cloud Composer, which is managed Airflow service by Google Cloud

### Deployment

Composer

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by ephraimbuddy** [comment]
Can you test this in Airflow 2.3.0

**Comment by nvmcrypto** [comment]
I'm having the same problem, I got the error sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) connection to server at "airflow-sqlproxy-service.default.svc.cluster.local" (10.67.240.74), port 3306 failed: server closed the connection unexpectedly This probably means the server terminated abnormally before or while processing the request. I'm using Google composer with the most up-to-date version.

**Comment by potiuk** [comment]
You need to take a look at your deployment logs in composer. It looks like your database connection was broken - and it could be for example because the firewall cut the connection or evicted the sql proxy. Many reasons, but this is not an airflow issue I am afraid.

**Reply by ilbus** [reply]
Please double check node pool used by Cloud Composer has proper service account, in my case during node auto-upgrades it was set to default - maybe some glitch - hard to say.
I have recreated default pool with gcloud commands  and now this heartbeat exception is gone

**Comment by aenni0409** [comment]
got the same issue here. Setup the airflow server via composer with airflow version 2.1.4 and composer 1.17.7 back to 02/2022. Everything is good understand recently. This error started shows up the end of April and became really bad on around 5/8. 
Creating a new composer did not work to me.

**Reply by potiuk** [reply]
I think you shoudl raise through Composer issue system - the error indicates this is a deployment issue, not Airflow.

**Comment by amalkumar** [comment]
@kavya-5 @danielpetrato @aenni0409 
Have any of you been able to resolve this issue?

I am also facing the same issue ( `composer-1.18.8 & airflow-2.2.3` ) but I have verified that the Cloud Composer node pool has a proper service account and the node pool service account is not set to default.

**Comment by naimsahin0** [comment]


**Reply by amalkumar** [reply]
@naimsahin0 

Thanks for your response.

We are now waiting for Google to come back on the issue in order to investigate the root cause, before trying the next version, i.e. composer-1.18.9-airflow-2.2.5.

**Reply by aenni0409** [reply]
@amalkumar, same here. Waiting for GCP

**Comment by ilbus** [comment]
Please double check permission used by node default-pool: you can do it with [gcloud container node-pools describe](https://cloud.google.com/sdk/gcloud/reference/container/node-pools/describe)
```
gcloud container node-pools describe default-pool \
--cluster=<<put your cloud composer cluster name>> \
--zone=<<put cluster location>>
```
Please check service account used there and have a look on oauth scopes.

I my case  service account was not defined so what I did I have recreated deafult-pool used by Cloud Composer
to do so:

1. Pause all DAGs
2. Note all properties of default-pool like, image version, machine_types, number of worker and other flags and settings 
3. use [gcloud container node-pools delete default-pool](https://cloud.google.com/sdk/gcloud/reference/container/node-pools/delete)
4. and recreate default pool with proper service account and scopes: [gcloud container node-pools create default-pool](https://cloud.google.com/sdk/gcloud/reference/container/node-pools/create) please reuse noted properties defined in step 2 and don't forget to define proper [service account](https://cloud.google.com/sdk/gcloud/reference/container/node-pools/create#--service-account)

**Reply by amalkumar** [reply]
@ilbus 

Thanks for your response.

In my case the  permissions(service account and oauth scopes) used by node default-pool looks fine.

**Comment by harshitasingh1612** [comment]
My airflow DAG was running file until morning today and suddenly it started throwing this error**: psycopg2.OperationalError: connection to server at "airflow-sqlproxy-service.default.svc.cluster.local" (10.16.6.20), port 3306 failed: server closed the connection unexpectedly.This probably means the server terminated abnormally before or while processing the request**

As per recommendation I upgraded to Airflow 2.2.5 but still the same.Any leads would be much appreciated.

**Reply by aenni0409** [reply]
contact the GCP support. We need more ppl to bring their attentions about this. I have a pending case with them currently and nothing change their status page

**Comment by filipknapik** [comment]
Hi all, I am a Product Manager for Cloud Composer. We are indeed seeing some cases with the error message above and our engineers are already investigating the issue.  I will update this thread as soon as have have a fix or a recommendation with next steps - it should happen soon as this is a priority case for us. Apologies for the inconvenience caused by the issue.

**Reply by mrrobby** [reply]
@filipknapik, any updates or temporary fixes for us? Can you please provide a link to the issue tracker?

**Reply by 4sushi** [reply]
Hello @filipknapik, we opened a ticket more than 10 days ago on Google Issue Tracker, and I still haven't had a single answer from Composer team. Today, we still have errors on several DAGs (`LocalTaskJob heartbeat got an exception`, `Lost connection to MySQL server at 'reading initial communication packet'`). You don't communicate about it. Here we are not talking about a problem with an open source library, but a cloud solution service that customers pay quite a lot for. You have to go and explore a Github discussion to finally find information from the Composer team. 

Ticker URL: https://issuetracker.google.com/issues/234038922 

Thanks.

**Reply by rohitsupekar** [reply]
Hello, any updates on this? We are facing similar issues and many of our production systems are experiencing difficulties.

**Comment by vishalgupt** [comment]
hi all,
I am using airflow 2.1.4 version and Cloud composer 1.17.4.  I am also getting same error - LocalTaskJob heartbeat got an exception and sqlalchemy.exc.OperationalError: (psycopg2.OperationalError).

**Comment by sanimesa** [comment]
Hello,

We started getting this error recently on Composer 1.18.7 Airflow 2.1.4. Happening randomly, even on EmailOperator tasks. Server resources are not at all stretched. 

`Failed to execute task (psycopg2.OperationalError) connection to server at "airflow-sqlproxy-service.default.svc.cluster.local" (10.79.242.18), port 3306 failed: server closed the connection unexpectedly
`

Please let us know if there are workarounds, maint that can be done to mitigate.

Thanks.

**Comment by 4sushi** [comment]
Hello,

Same problem here, we have random errors on all DAGs in all of our Composer instances (`composer-1.18.0-airflow-1.10.15`)  : 
```
Lost connection to MySQL server at 'reading initial communication packet', system error: 0
```

I had open an issue here: [https://issuetracker.google.com/issues/234038922](https://issuetracker.google.com/issues/234038922)
Composer support is not responsive.

**Comment by bthng-poly** [comment]
Do we know what is causing this issue? I face this almost daily, and oddly it only happens for a set of dags.

**Reply by potiuk** [reply]
Some connectivity issue. But what - only you can tell after looking at your networking/configuration etc.

**Reply by bthng-poly** [reply]
Update:
Just in case it helps someone, I checked the logs and seems like our composer setup was under-resourced and hence is causing connection errors. Go to the `Environment details` page --> `view cluster workloads`. You'll be able to check on whether the composer setup is sufficiently resourced. In our case, we had to increase the number of workers and schedulers.

**Comment by radirobi97** [comment]
Hello everybody,

we have just received an update. The problem could be related to the sql-proxy. This issue happens when sql-proxy lacks of CPU resources, so try to increase resource limits/requests of your sql-proxy.

**Reply by kieronellis** [reply]
> 

Hi [radirobi97](https://github.com/radirobi97), how exactly do you increase the resource limits/requests of the sql-proxy? Thanks

**Reply by radirobi97** [reply]
@kieronellis we are using Cloud Composer which is a managed product by Google, basically a wrapper around Airflow with many features. In Composer, Airflow is running on top of a Kubernetes cluster, so the proper yaml file of the proxy should be adjusted in the resource request section.

**Reply by rohitsupekar** [reply]
Thanks so much for the update! Would you be able to paste what exactly you modified in the yaml file?

**Reply by radirobi97** [reply]
```yaml
spec:
  template:
    spec:
      containers:
        - name: airflow-sqlproxy
           resources:
             requests:
               cpu: 100m
```

**Comment by filipknapik** [comment, marked as answer]
The fix for this issue is available in Cloud Composer 1.19.0, which has just started its rollout. The new version will start showing up in individual regions gradually until Monday. 

You can upgrade your Composer 1 environments where you experience this issue to version 1.19.0 to acquire the fix immediately.
The fix will also be applied to all existing Composer 1 environments within the next few days and so upgrading is not necessary if you are OK to wait for it. 

We will announce when all existing environments get the fix in [Cloud Composer Release Notes](https://cloud.google.com/composer/docs/release-notes). If you have any questions or concerns, feel free to contact [Cloud Customer Care](https://cloud.google.com/support).

Again, I apologize for the inconvenience caused by this issue. On top of applying the fix for this issue, we are also performing an extensive root cause analysis aiming to prevent similar issues from happening ever again.
